stan/math/prim/meta/is_matrix_cl.hpp

```diff
@@ -0,0 +1,29 @@
+#ifndef STAN_MATH_PRIM_META_IS_MATRIX_CL_HPP
+#define STAN_MATH_PRIM_META_IS_MATRIX_CL_HPP
+
+#include <stan/math/prim/meta/require_helpers.hpp>
+#include <type_traits>
+
+namespace stan {
+namespace math {
+
+/**
+ * Non-templated base class for `matrix_cl` simplifies checking if something is
+ * matrix_cl.
+ */
+class matrix_cl_base {};
+
+}  // namespace math
+
+/** \ingroup type_traits
+ * Checks if the decayed type of T is a matrix_cl.
+ */
+template <typename T>
+struct is_matrix_cl
+    : public std::is_base_of<math::matrix_cl_base, std::decay_t<T>> {};
```

Comment:
What's the logic here on doing a base class versus an impl method like we do for `is_std_vector`?

Reply:
This also works for classes derived from `matrix_cl` (`arena_matrix_cl`). I could also add a new case to previous implementation, but this is less code.